drivers/sensors/pmw3389.c

```diff
@@ -0,0 +1,288 @@
+
+// clang-format on
+
+#ifndef MAX_CPI
+#    define MAX_CPI 0x0140  // limits to 0--320, should be max cpi/50
+#endif
+
+bool _inBurst = false;
+
+#ifdef CONSOLE_ENABLE
+void print_byte(uint8_t byte) { dprintf("%c%c%c%c%c%c%c%c|", (byte & 0x80 ? '1' : '0'), (byte & 0x40 ? '1' : '0'), (byte & 0x20 ? '1' : '0'), (byte & 0x10 ? '1' : '0'), (byte & 0x08 ? '1' : '0'), (byte & 0x04 ? '1' : '0'), (byte & 0x02 ? '1' : '0'), (byte & 0x01 ? '1' : '0')); }
+#endif
+#define constrain(amt, low, high) ((amt) < (low) ? (low) : ((amt) > (high) ? (high) : (amt)))
+
+bool spi_start_adv(void) {
+    bool status = spi_start(PMW3389_CS_PIN, PMW3389_SPI_LSBFIRST, PMW3389_SPI_MODE, PMW3389_SPI_DIVISOR);
+    wait_us(1);
+    return status;
+}
+
+void spi_stop_adv(void) {
+    wait_us(1);
+    spi_stop();
+}
+
+spi_status_t spi_write_adv(uint8_t reg_addr, uint8_t data) {
+    if (reg_addr != REG_Motion_Burst) {
+        _inBurst = false;
+    }
+
+    spi_start_adv();
+    // send address of the register, with MSBit = 1 to indicate it's a write
+    spi_status_t status = spi_write(reg_addr | 0x80);
+    status              = spi_write(data);
+
+    // tSCLK-NCS for write operation
+    wait_us(20);
+
+    // tSWW/tSWR (=120us) minus tSCLK-NCS. Could be shortened, but is looks like a safe lower bound
+    wait_us(100);
+    spi_stop();
+    return status;
+}
+
+uint8_t spi_read_adv(uint8_t reg_addr) {
+    spi_start_adv();
+    // send adress of the register, with MSBit = 0 to indicate it's a read
+    spi_write(reg_addr & 0x7f);
+
+    uint8_t data = spi_read();
+
+    // tSCLK-NCS for read operation is 120ns
+    wait_us(1);
+
+    //  tSRW/tSRR (=20us) minus tSCLK-NCS
+    wait_us(19);
+
+    spi_stop();
+    return data;
+}
+
+void pmw3389_set_cpi(uint16_t cpi) {
+    uint16_t cpival = constrain((cpi / 50) - 1, 0, MAX_CPI);
+
+    spi_start_adv();
+    spi_write_adv(REG_Resolution_L, cpival & 0xFF);
+    spi_write_adv(REG_Resolution_H, (cpival >> 8) & 0xFF);
+    spi_stop();
+}
+
+uint16_t pmw3389_get_cpi(void) {
+    uint16_t cpival = (spi_read_adv(REG_Resolution_H) << 8) | spi_read_adv(REG_Resolution_L);
+    return (uint16_t)((cpival + 1) & 0xFFFF) * 50;
+}
+
+bool pmw3389_init(void) {
+    setPinOutput(PMW3389_CS_PIN);
+
+    spi_init();
+    _inBurst = false;
+
+    spi_stop();
+    spi_start_adv();
+    spi_stop();
+
+    spi_write_adv(REG_Shutdown, 0xb6);  // Shutdown first
+    wait_ms(300);
+
+    spi_start_adv();
+    wait_us(40);
+    spi_stop_adv();
+    wait_us(40);
+
+    spi_write_adv(REG_Power_Up_Reset, 0x5a);
+    wait_ms(50);
+
+    spi_read_adv(REG_Motion);
+    spi_read_adv(REG_Delta_X_L);
+    spi_read_adv(REG_Delta_X_H);
+    spi_read_adv(REG_Delta_Y_L);
+    spi_read_adv(REG_Delta_Y_H);
+
+    pmw3389_upload_firmware();
+
+    spi_stop_adv();
+
+    wait_ms(10);
+    pmw3389_set_cpi(PMW3389_CPI);
+
+    wait_ms(1);
+
+    spi_write_adv(REG_Config2, 0x00);
+
+    spi_write_adv(REG_Angle_Tune, constrain(ROTATIONAL_TRANSFORM_ANGLE, -30, 30));
```

Comment:
This should be changed to -127 to 127. 
```suggestion
    spi_write_adv(REG_Angle_Tune, constrain(ROTATIONAL_TRANSFORM_ANGLE, -127, 127));
```

Reply:
hhhhmmm I didn't see anything in the docs suggesting that changed and from the 3360 docs it is clearly -30 to 30
![image](https://user-images.githubusercontent.com/96358682/146666498-1935321a-144d-49c8-bce4-dfc881081114.png)

Reply:
The same info is in the 3360 datasheet.  But I can verify that it allows the full int8_t range for this register on the 3360, and I'm guessing that it's the same here.

Reply:
interesting I wonder what happens to values above 30 and below -30 as based on the hex values shown in that sheet they are unsupported (maybe just defaults to 30 and -30)
by the way based on those hex values it is expecting the two's complement of the negative values do we need any special handling for that?

Reply:
On the 3360, it "just works", and respects the higher values. 

And no, nothing needs to be done to convert it.

Reply:
huh rereading the documentation nowhere is it stated that 30 and -30 are the maximums...
Alright change is committed as requested